File: epi_judge_java/epi/RemoveDuplicatesFromSortedList.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

public class RemoveDuplicatesFromSortedList {

    @EpiTest(testDataFile = "remove_duplicates_from_sorted_list.tsv")
    public static ListNode<Integer> removeDuplicates(ListNode<Integer> head) {
        if (head == null) return null;

        var curr = head;
        var next = head.next;

        while (next != null) {
            if (!curr.data.equals(next.data)) {
                curr.next = next;
                curr = curr.next;
            }

            next = next.next;
        }

        curr.next = null;

        return head;
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "RemoveDuplicatesFromSortedList.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/remove_duplicates_from_sorted_list.cc
#include <memory>

#include "list_node.h"
#include "test_framework/generic_test.h"

using std::shared_ptr;

shared_ptr<ListNode<int>> RemoveDuplicates(const shared_ptr<ListNode<int>> &head) {
    if (!head) return head;

    auto read = head->next;
    auto write = head;

    while (read) {
        if (read->data != write->data) {
            write->next = read;
            write = write->next;
        }

        read = read->next;
    }

    write->next = nullptr;

    return head;
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"L"};
    return GenericTestMain(args, "remove_duplicates_from_sorted_list.cc",
                           "remove_duplicates_from_sorted_list.tsv",
                           &RemoveDuplicates, DefaultComparator{}, param_names);
}
